src/effects/effects_base/internal/effectsprovider.cpp

```diff
@@ -365,45 +338,41 @@ static void EffectPreview(EffectBase& effect,
 
         access.ModifySettings([&](EffectSettings& settings){
             // Preview of non-realtime effect
-            auto pInstance
-                =std::dynamic_pointer_cast<EffectInstanceEx>(effect.MakeInstance());
+            auto pInstance = std::dynamic_pointer_cast<EffectInstanceEx>(effect.MakeInstance());
             success = pInstance && pInstance->Process(settings);
             return nullptr;
         });
     }
 
     if (success) {
-        auto tracks = MakeTransportTracks(*mTracks, true);
-
         // Some effects (Paulstretch) may need to generate more
         // than previewLen, so take the min.
         t1 = std::min(mT0 + previewLen, mT1);
 
         // Start audio playing
-        auto options = ProjectAudioIO::GetDefaultOptions(*pProject);
-        int token = gAudioIO->StartStream(tracks, mT0, t1, t1, options);
+        playback::PlayTracksOptions opt = { .selectedOnly = true,
+                                            .startOffset = startOffset };
+        muse::Ret ret = player->playTracks(*mTracks, mT0, t1, opt);
 
-        if (token) {
+        if (ret) {
             using namespace BasicUI;
             auto previewing = BasicUI::ProgressResult::Success;
             // The progress dialog must be deleted before stopping the stream
             // to allow events to flow to the app during StopStream processing.
             // The progress dialog blocks these events.
             {
-                auto progress = MakeProgress(effect.GetName(),
-                                             XO("Previewing"), ProgressShowStop);
+                auto progress = MakeProgress(effect.GetName(), XO("Previewing"), ProgressShowStop);
 
-                while (gAudioIO->IsStreamActive(token) && previewing == BasicUI::ProgressResult::Success) {
+                while (player->isRunning() && previewing == BasicUI::ProgressResult::Success) {
```

Comment:
Thank you for telling me about the meaning of the token. In AU4 this will be implemented a little differently, each instance will have its own player object. Therefore, here, when using the player, we do not need to think about the token, but in the player itself the token is now stored through the use of `ProjectAudioIO::Get(project).SetAudioIOToken(token);`